### PR TITLE
Fix internal flags not always being written when they should be

### PR DESCRIPTION
**Pull request by Sushisource**
## What was changed
Internal flags weren't always written in the first WFT completion where you might actually care about that. Rather than littering the code with a bunch of accesses just to make sure that it always shows up where needed, given these are just ints, I'm going back on what I thought earlier where writing them lazily is best. Seems totally fine to just spit out all the known ones on the first non-replay WFT.

Also restructured the flags structure a bit to be somewhat safer.

## Why?
Right now if you, ex, have a workflow with one activity which sched/starts/completes in one WFT, determinism type checking will not work for it, since the flag wasn't written until the scheduled event is first seen.

## Checklist

1. Closes 

2. How was this tested:
Added tests, changed existing tests

3. Any docs updates needed?

### CODE REVIEW

**Comment by cretz** [comment]
Mind updating api_upstream as part of this to get https://github.com/temporalio/api/pull/270?

**Comment by Sushisource** [comment]
> Mind updating api_upstream as part of this to get [temporalio/api#270](https://github.com/temporalio/api/pull/270)?

Done

File: core/src/internal_flags.rs
Comment (mjameswh) [inline core/src/internal_flags.rs:92]: I don't understand the reasoning here. Shouldn't we check if the flag already exists _first_? As it is now, a patch might be registered multiple time...

Comment (Sushisource) [reply core/src/internal_flags.rs:92]: The dedupe just happens on the way out, rather than the way in - in `gather_for_wft_complete`

Comment (mjameswh) [reply core/src/internal_flags.rs:92]: Clever. Might be worth adding a few comments as this is definitely not obvious, but not blocking this PR.

**Review by mjameswh — APPROVED** [review]
LGTM

**Review by bergundy — APPROVED** [review]
Didn't review, just stamping to get codeowner approval.
@mjameswh already reviewed this.

**Pull request merged**
